platform/linux-generic/pktio/mdev.c

```diff
@@ -0,0 +1,585 @@
+/* Copyright (c) 2018, Linaro Limited
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier:     BSD-3-Clause
+ */
+
+#include "config.h"
+
+#ifdef ODP_MDEV
+
+#include <odp_posix_extensions.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/ioctl.h>
+#include <unistd.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <linux/vfio.h>
+#include <stddef.h>
+#include <stdlib.h>
+#include <dirent.h>
+
+#include <config.h>
+#include <odp_posix_extensions.h>
+#include <odp/api/hints.h>
+
+#include <pktio/mdev.h>
+#include <pktio/uapi_net_mdev.h>
+
+#include <odp_debug_internal.h>
+#include <odp_align_internal.h>
+
+/**
+ * returns a valid VFIO container
+ * fd must be closed by caller
+ */
+static int get_container(void)
+{
+	int ret;
+	int container;
+	/* Create a new container */
+	container = open("/dev/vfio/vfio", O_RDWR);
+
+	if (container < 0)
+		return container;
+
+	ret = ioctl(container, VFIO_GET_API_VERSION);
+	if (ret != VFIO_API_VERSION) {
+		ODP_ERR("Unknown API version\n");
+		goto out;
+	}
+
+	if (!ioctl(container, VFIO_CHECK_EXTENSION, VFIO_TYPE1_IOMMU)) {
+		ODP_ERR("Doesn't support the IOMMU driver we want\n");
+		goto out;
+	}
+
+	return container;
+out:
+	close(container);
+	container = -1;
+	return ret;
+}
+
+/**
+ * returns a valid VFIO group
+ * fd must be close by caller
+ */
+static int get_group(int grp_id)
+{
+	char path[64];
+	int ret;
+	int group;
+	struct vfio_group_status group_status = {
+	    .argsz = sizeof(group_status)
+	};
+
+	snprintf(path, sizeof(path), "/dev/vfio/%d", grp_id);
+	group = open(path, O_RDWR);
+	if (group < 0) {
+		ODP_ERR("Failed to open %s, %d (%s)\n",
+			path, group, strerror(errno));
+		return group;
+	}
+
+	ret = ioctl(group, VFIO_GROUP_GET_STATUS, &group_status);
+
+	if (ret < 0) {
+		ODP_ERR("ioctl(VFIO_GROUP_GET_STATUS) failed\n");
+		goto out;
+	}
+
+	/* Test the group is viable and available */
+	if (!(group_status.flags & VFIO_GROUP_FLAGS_VIABLE)) {
+		ODP_ERR("Group is not viable\n");
+		goto out;
+	}
+
+	return group;
+out:
+	close(group);
+	group = -1;
+	return ret;
```

Comment:
Negative to indicate error and 0 and positive to indicate success?

Reply:
Thanks, fixed.